File: xkcd/src/main/java/xyz/jienan/xkcd/model/util/WhatIfArticleUtil.java
package xyz.jienan.xkcd.model.util;

import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;
import org.jsoup.select.Elements;

import java.io.IOException;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.List;
import java.util.Locale;

import okhttp3.ResponseBody;
import xyz.jienan.xkcd.model.WhatIfArticle;

public class WhatIfArticleUtil {

    private static final String BASE_URI = "https://what-if.xkcd.com/";

    private static final String TEX_JS = "https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.4/latest.js?config=TeX-MML-AM_CHTML";

    public static List<WhatIfArticle> getArticlesFromArchive(ResponseBody responseBody) throws IOException, ParseException {
        Document doc = Jsoup.parse(responseBody.string(), BASE_URI);
        Elements divArchive = doc.select("div#archive-wrapper");
        List<WhatIfArticle> articles = new ArrayList<>();
        for (Element element : divArchive.first().children()) {
            WhatIfArticle article = new WhatIfArticle();

            Element a = element.selectFirst("a[href]");
            String[] href = a.attr("href").split("/");
            article.num = Long.parseLong(href[href.length - 1]);
            article.featureImg = a.child(0).absUrl("src");
            String archiveDate = element.selectFirst("h2.archive-date").html();
            final SimpleDateFormat sdf = new SimpleDateFormat("MMMMM d, yyyy", Locale.ENGLISH);
            article.date = sdf.parse(archiveDate).getTime();

            articles.add(article);
        }
        return articles;
    }

    public static Document getArticleFromHtml(ResponseBody responseBody) throws IOException {
        final Document doc = Jsoup.parse(responseBody.string(), BASE_URI);
        final Elements elements = doc.select("article.entry");
        elements.remove(elements.get(0).children().first());

        final Elements imageElements = doc.select("img.illustration");
        for (Element element : imageElements) {
            element.attr("src", element.absUrl("src"));
        }

        final Elements pElements = doc.select("p");
        for (Element element : pElements) {
            if (element.html().split("\\[").length > 1) {
                element.attr("class", "latex");
            }
        }

        final Elements aElements = doc.select("a");
        for (Element element : aElements) {
            element.attr("href", element.absUrl("href"));
        }

        doc.head().html("");
        doc.head().appendElement("link")
                .attr("rel", "stylesheet")
                .attr("type", "text/css")
                .attr("href", "style.css");
        doc.head().appendElement("script")
                .attr("src", TEX_JS)
                .attr("async", "");
        doc.head().appendElement("script").attr("src", "LatexInterface.js");
        doc.head().appendElement("script").attr("src", "ImgInterface.js");
        doc.head().appendElement("script").attr("src", "RefInterface.js");
        doc.body().html(elements.html()).appendElement("p");
        return doc;
    }
}

File: xkcd/src/main/java/xyz/jienan/xkcd/model/util/WhatIfArticleUtil.kt
package xyz.jienan.xkcd.model.util

import okhttp3.ResponseBody
import org.jsoup.Jsoup
import org.jsoup.nodes.Document
import org.jsoup.nodes.Element
import timber.log.Timber
import xyz.jienan.xkcd.model.WhatIfArticle
import java.io.IOException
import java.text.ParseException
import java.text.SimpleDateFormat
import java.util.*

object WhatIfArticleUtil {

    private const val BASE_URI = "https://what-if.xkcd.com/"

    private const val TEX_JS = "https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.4/latest.js?config=TeX-MML-AM_CHTML"

    @Throws(IOException::class, ParseException::class)
    fun getArticlesFromArchive(responseBody: ResponseBody): List<WhatIfArticle> {
        val doc = Jsoup.parse(responseBody.string(), BASE_URI)
        val articleDivs = doc.select("div.archive-entry")

        Timber.d("Article divs size ${articleDivs.size}")
        return articleDivs.map { it.getWhatIfArticleFromArchiveEntry() }
    }

    @Throws(IOException::class)
    fun getArticleFromHtml(responseBody: ResponseBody): Document {
        val doc = Jsoup.parse(responseBody.string(), BASE_URI)

        doc.head().appendScriptsToHead()

        doc.selectFirst("article#entry").apply {
            cleanUp()
            appendElement("p")
            doc.body().html(html())
        }

        return doc
    }

    private fun Element.cleanUp() {
        select("img.illustration").forEach { it.convertToFullHttpsImgUrl() }
        select("p").filter { it.html().split("\\[").size > 1 }.forEach { it.tagLatex() }
        select("a").map { it.attr("href", it.absUrl("href")) }
    }

    private fun Element.convertToFullHttpsImgUrl() {
        attr("src", absUrl("src").replaceFirst("^http://what".toRegex(), "https://what"))
    }

    private fun Element.tagLatex() {
        when {
            html().startsWith("\\[") -> attr("class", "latex")
            parent().html().contains("<a href=\"//what-if.xkcd.com/124/\">") -> {
                val contents = html()
                contents.replace("\\(", "\\[")
                contents.replace("\\)", "\\]")
                html(contents).appendElement("br")
            }
            else -> {
                val taggedLatex = html().replace("(?=\\\\\\[)".toRegex(), "<p class=\"latex\">")
                        .replace("(?<=\\\\\\])".toRegex(), "<p>")
                html(taggedLatex)
            }
        }
    }

    private fun Element.appendScriptsToHead() {
        empty()
        appendCss("style.css")
        appendElement("script")
                .attr("src", TEX_JS)
                .attr("async", "")
        appendElement("script").attr("src", "LatexInterface.js")
        appendElement("script").attr("src", "ImgInterface.js")
        appendElement("script").attr("src", "RefInterface.js")
    }

    private fun Element.getWhatIfArticleFromArchiveEntry(): WhatIfArticle {
        val a = selectFirst("a[href]")
        val href = a.attr("href").split("/".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
        val archiveDate = selectFirst("h3.archive-date").html()
        val sdf = SimpleDateFormat("MMMMM d, yyyy", Locale.ENGLISH)
        val title = selectFirst("h2.archive-title a").html()
        val num = href.last().toLong()
        val featureImg = a.child(0).absUrl("src")

        return WhatIfArticle(num = num,
                title = title,
                featureImg = featureImg,
                date = sdf.parse(archiveDate).time)
    }
}

fun Element.appendCss(cssName: String) {
    appendElement("link")
            .attr("rel", "stylesheet")
            .attr("type", "text/css")
            .attr("href", cssName)
}
